### DISCUSSION TITLE
Request to /api folder from getInitialProps

### DISCUSSION THREAD

**Discussion by oran1248** [vercel/next.js#18201] [Help]
Hi,

In my app, I'm sending request from `getInitialProps`  to an `api` endpoint (in `/api` folder).
When running my app in localhost, everything works, because it's all under same domain (`localhost`).

After deploying my app successfully to aws-lambda (using `serverless-next.js`), those requests doesn't work anymore.
The reason is that the host header when running `getInitialProps` in server, is `myAppBucket.s3.us-east-1.amazonaws.com`, but the app domain is `https://d25q7fh11ll2cg.cloudfront.net.`.
request to `myAppBucket.s3.us-east-1.amazonaws.com/api/users` doesn't work but `https://d25q7fh11ll2cg.cloudfront.net/api/users` works. 
But I don't have `https://d25q7fh11ll2cg.cloudfront.net/api/users` when running getInitialProps.

I don't want to put the domain as an env var. there is another solution for this?
Thanks

**Comment by ats1999** [comment]
Don't call api routes in getinitialprops. Instead you can make a hander function and call it from getinitialprops.

**Reply by oran1248** [reply]
@ats1999 
Do you have an example? seems weird to me calling directly the handler. because `/api` is server-side code and `getInitialProps` is both client side and server side. How it will work exactly? Thanks :)

EDIT: I tried it, but i'm getting error because my handler use a server-only packages, but webpack is trying to bundle it because it appears in getInitialProps.

**Reply by ats1999** [reply]
> @ats1999
> Do you have an example? seems weird to me calling directly the handler. because `/api` is server-side code and `getInitialProps` is both client side and server side. How it will work exactly? Thanks :)
> 
> EDIT: I tried it, but i'm getting error because my handler use server-only packages, but webpack is trying to bundle it because it appears in getInitialProps.

`getInitialProps` is meant for the client-side, but it will be always executed on the server-side. So, you can call server-side code directly in `getInitialProps`. This includes reading from the filesystem or a database.

> Fetching from an external API is fine!

**Reply by oran1248** [reply]
@ats1999 
But in my handler I use server-only packages like 'mongodb' so im getting webpack error when trying to build the app because in my packages.json I set mongodb: false in the browser section so it will no be bundled and sent to the client. Should i solve this with dynamic imports at runtime in my handler and not static?  I will not hurt app's performance?

**Comment by oran1248** [comment]
@timneutkens 
please help, it seems to me a very common use case, but I can't find a good solution.

**Reply by ats1999** [reply]
As NextJS suggest you should not call APi directly from server-side rendering.
>Note: You should not use fetch() to call an API route in your application. Instead, directly import the API route and call its function yourself. You may need to slightly refactor your code for this approach.

[Check out](https://nextjs.org/docs/basic-features/data-fetching)
